### ISSUE TITLE
uneven io distribution

### ISSUE THREAD

**Issue by n-other** [zfsonlinux/zfs#3686]
Hi,

while resolving this issue https://github.com/zfsonlinux/zfs/issues/3670 i've created additional dataset in the pool and started to copy with rsync data from old dataset to new dataset within the same pool.

the issue I see - there is extra load on the first five raidz-sets, while others are getting much less hit. here is iostat output from the pool.

```
                                                capacity     operations    bandwidth
pool                                         alloc   free   read  write   read  write
-------------------------------------------  -----  -----  -----  -----  -----  -----
msapool                                      85.5T  1.45T    574    370  66.7M  35.5M
  raidz1                                     7.18T  74.1G     49     55  5.91M  5.31M
    c0d3                                         -      -     23     25  1.58M  1.84M
    c1d3                                         -      -     20     23  1.53M  1.84M
    c2d3                                         -      -     23     25  1.57M  1.84M
    c3d3                                         -      -     20     22  1.52M  1.84M
  raidz1                                     7.18T  74.1G     52     57  6.15M  5.54M
    c0d4                                         -      -     24     30  1.58M  1.85M
    c1d4                                         -      -     25     29  1.58M  1.85M
    c2d4                                         -      -     25     30  1.58M  1.85M
    c3d4                                         -      -     25     29  1.58M  1.85M
  raidz1                                     7.18T  74.0G     99     52  12.0M  4.87M
    c0d2                                         -      -     55     31  4.01M  1.63M
    c1d2                                         -      -      0     64  5.34K  3.96M
    c2d2                                         -      -     54     31  4.01M  1.63M
    c3d2                                         -      -     56     30  4.01M  1.63M
  raidz1                                     7.18T  74.2G     98     46  11.8M  4.70M
    c0d1                                         -      -     54     26  3.96M  1.57M
    c1d1                                         -      -     54     26  3.96M  1.57M
    c2d1                                         -      -      0     60  5.24K  3.90M
    c3d1                                         -      -     56     26  3.96M  1.57M
  raidz1                                     7.18T  75.2G     97     48  11.9M  5.07M
    c0d0                                         -      -      0     58  33.9K  3.91M
    c1d0                                         -      -     54     26  3.97M  1.69M
    c2d0                                         -      -     54     26  3.97M  1.69M
    c3d0                                         -      -     56     26  3.97M  1.69M
  raidz1                                     7.18T  76.7G     17     19  1.74M  1.57M
    c0d5                                         -      -      7     12   453K   537K
    c1d5                                         -      -      7     12   453K   536K
    c2d5                                         -      -      7     12   453K   537K
    c3d5                                         -      -      8     12   453K   536K
  raidz1                                     7.17T  76.9G     16     18  1.74M  1.56M
    c0d6                                         -      -      7     11   453K   535K
    c1d6                                         -      -      7     11   453K   534K
    c2d6                                         -      -      7     11   453K   535K
    c3d6                                         -      -      7     11   453K   535K
  raidz1                                     7.18T  76.8G     17     17  1.88M  1.68M
    c0d7                                         -      -      8     10   490K   575K
    c1d7                                         -      -      8     10   490K   575K
    c2d7                                         -      -      8     10   490K   575K
    c3d7                                         -      -      8     10   490K   575K
  raidz1                                     7.17T  86.8G     80     23  9.77M  2.61M
    c0d8                                         -      -     45     12  3.27M   893K
    cciss-3600508b30084adf0a051c23434310033      -      -      0     45  1.40K  3.21M
    c2d8                                         -      -     44     12  3.26M   893K
    c3d8                                         -      -     45     12  3.26M   893K
  raidz1                                     6.91T   351G     10      6  1.02M   616K
    c0d9                                         -      -      4      3   262K   206K
    c1d9                                         -      -      4      3   261K   206K
    c2d9                                         -      -      4      3   262K   206K
    c3d9                                         -      -      4      3   261K   206K
  raidz1                                     7.00T   252G     20      9  1.39M   711K
    c0d10                                        -      -      8      6   361K   238K
    c1d10                                        -      -      7      6   361K   237K
    c2d10                                        -      -      7      6   361K   238K
    c3d10                                        -      -      7      6   361K   237K
  raidz1                                     7.06T   193G     14     14  1.39M  1.26M
    c0d11                                        -      -      6      7   362K   431K
    c1d11                                        -      -      6      7   362K   431K
    c2d11                                        -      -      6      7   362K   431K
    c3d11                                        -      -      6      7   362K   431K
cache                                            -      -      -      -      -      -
  da49                                           -      -      0      0      0      0
  da50                                           -      -      0      0      0      0
  da51                                           -      -      0      0      0      0
  da48                                           -      -      0      0      0      0
-------------------------------------------  -----  -----  -----  -----  -----  -----
```

**Comment by behlendorf** [comment]
@n-other did the load even out over time?  I notice your pool is very full which always complicates allocation policy.

**Comment by n-other** [comment]
@behlendorf it did. rsync copy lasts for the last 6 days. first 4 days I observed the same as above distribution, however for the last 2 days it looks like this:

zpool iostat -v|grep raid
  raidz1                                     7.18T  74.0G     41     45  4.90M  4.40M
  raidz1                                     7.18T  74.1G     43     47  5.08M  4.56M
  raidz1                                     7.18T  73.8G     83     46  10.1M  4.38M
  raidz1                                     7.18T  73.9G     84     42  10.1M  4.35M
  raidz1                                     7.18T  74.0G     81     42  9.93M  4.45M
  raidz1                                     7.17T  78.8G     32     36  3.67M  3.35M
  raidz1                                     7.17T  79.4G     32     35  3.67M  3.35M
  raidz1                                     7.18T  76.3G     32     34  3.71M  3.38M
  raidz1                                     7.18T  73.8G     74     30  8.93M  3.20M
  raidz1                                     7.18T  74.8G     27     31  3.20M  3.07M
  raidz1                                     6.49T   774G     44     35  4.34M  3.44M
  raidz1                                     6.68T   587G     36     37  4.09M  3.70M

So, IO writes are more or less distributed across all raids groups. However there is still something interesting going on with space allocation. In the first output, 3 last raidz groups are having more free space than others (351 + 252 + 193 = 796G), however current snapshot shows only last 2 raidz groups having more space than others (774 + 587 = 1371G). I've deleted 503G of data from the dataset yesterday, so it's clear why it has jumped from 796 to 1371, but still it does not explain why it has allocated like this
